lib/hanami/slice.rb

```diff
@@ -205,27 +304,114 @@ def prepare_container_component_dirs # rubocop:disable Metrics/AbcSize
 
         # TODO: Change `""` (signifying the root) once dry-rb/dry-system#238 is resolved
         container.config.component_dirs.add("") do |dir|
+          # TODO: ignore lib/ child dir here
           dir.namespaces.add_root(key: nil, const: slice_name.name)
           dir.auto_register = auto_register_proc.(root)
         end
       end
 
+      def prepare_container_imports
+        import(
+          keys: config.slices.shared_component_keys,
+          from: application.container,
+          as: nil
+        )
+      end
+
+      def prepare_container_providers
+        # Check here for the `routes` definition only, not `router` itself, because the
+        # `router` requires the slice to be prepared before it can be loaded, and at this
+        # point we're still in the process of preparing.
+        if routes
+          require_relative "providers/routes"
+          register_provider(:routes, source: Hanami::Providers::Routes.for_slice(self))
+        end
+
+        if settings
+          require_relative "providers/settings"
+          register_provider(:settings, source: Hanami::Providers::Settings.for_slice(self))
+        end
+      end
```

Comment:
I just want to point out this out because I'm particularly chuffed about this outcome: if we hadn't added support for provider sources as concrete classes to dry-system, there's no way we could've set up this kind of conditional provider loading against multiple arbitrary containers! ✨